File: app/src/main/java/com/hkb48/keepdo/widget/TasksWidgetModel.java
package com.hkb48.keepdo.widget;

import android.content.Context;
import android.util.Log;

import com.hkb48.keepdo.DatabaseAdapter;
import com.hkb48.keepdo.Task;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

class TasksWidgetModel {
    private static final String TAG = "#TasksWidgetModel: ";
    private static final String SDF_PATTERN_YMD = "yyyy-MM-dd";

    private final Context mContext;
    private List<Task> mTaskList = new ArrayList<>();

    TasksWidgetModel (Context context) {
        mContext = context;
    }

    void reload() {
        mTaskList.clear();
        List<Task> fullTaskList = DatabaseAdapter.getInstance(mContext).getTaskList();
        String today = getTodayDate();
        for (Task task : fullTaskList) {
            if (! getDoneStatus(task.getTaskID(), today) && isValidDay(task, today)) {
                mTaskList.add(task);
            }
        }
    }

    int getItemCount() {
        return mTaskList.size();
    }

    long getTaskId(int position) {
        return mTaskList.get(position).getTaskID();
    }

    String getTaskName(int position) {
        return mTaskList.get(position).getName();
    }

    boolean getDoneStatus(long taskId, String date) {
        DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance(mContext);
        return dbAdapter.getDoneStatus(taskId, date);
    }

    String getTodayDate() {
        DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance(mContext);
        return dbAdapter.getTodayDate();
    }

    private boolean isValidDay(Task task, String dateString) {
        final SimpleDateFormat sdf_ymd = new SimpleDateFormat(SDF_PATTERN_YMD, Locale.JAPAN);
        Date date = null;
        if (dateString != null) {
            try {
                date = sdf_ymd.parse(dateString);
            } catch (ParseException e) {
                Log.e(TAG, e.getMessage());
                return false;
            }
        }

        if (date == null) {
            return false;
        }

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK);
        return task.getRecurrence().isValidDay(dayOfWeek);
    }
}

File: app/src/main/java/com/hkb48/keepdo/widget/TasksWidgetModel.kt
package com.hkb48.keepdo.widget

import android.content.Context
import com.hkb48.keepdo.DateChangeTimeUtil
import com.hkb48.keepdo.KeepdoApplication
import com.hkb48.keepdo.Recurrence
import com.hkb48.keepdo.db.entity.Task
import kotlinx.coroutines.runBlocking
import java.util.*

internal class TasksWidgetModel(private val mContext: Context) {
    private val mTaskList: MutableList<Task> = ArrayList()

    fun reload() = runBlocking {
        mTaskList.clear()
        val fullTaskList =
            (mContext as KeepdoApplication).getDatabase().taskDao().getTaskListByOrder()
        for (task in fullTaskList) {
            if (getDoneStatus(task._id!!, todayDate).not() && isValidDay(task, todayDate)) {
                mTaskList.add(task)
            }
        }
    }

    val itemCount: Int
        get() = mTaskList.size

    fun getTaskId(position: Int): Int {
        return mTaskList[position]._id!!
    }

    fun getTaskName(position: Int): String {
        return mTaskList[position].name
    }

    private suspend fun getDoneStatus(taskId: Int, date: Date): Boolean {
        return (mContext as KeepdoApplication).getDatabase().taskCompletionDao()
            .getByDate(taskId, date).count() > 0
    }

    val todayDate: Date
        get() = DateChangeTimeUtil.dateTime

    private fun isValidDay(task: Task, date: Date?): Boolean {
        return if (date != null) {
            val calendar = Calendar.getInstance()
            calendar.time = date
            Recurrence.getFromTask(task).isValidDay(calendar[Calendar.DAY_OF_WEEK])
        } else {
            false
        }
    }
}